pkg/sql/tenant.go

```diff
@@ -470,40 +470,79 @@ func clearTenant(ctx context.Context, execCfg *ExecutorConfig, info *descpb.Tena
 }
 
 // DestroyTenant implements the tree.TenantOperator interface.
-func (p *planner) DestroyTenant(ctx context.Context, tenID uint64, synchronous bool) error {
-	const op = "destroy"
-	if err := p.RequireAdminRole(ctx, "destroy tenant"); err != nil {
+func (p *planner) DestroyTenant(
+	ctx context.Context, tenantName roachpb.TenantName, synchronous bool,
+) error {
+	if err := p.validateDestroyTenant(ctx); err != nil {
 		return err
 	}
-	if err := rejectIfCantCoordinateMultiTenancy(p.execCfg.Codec, op); err != nil {
-		return err
+
+	info, err := GetTenantRecordByName(ctx, p.execCfg, p.txn, tenantName)
+	if err != nil {
+		return errors.Wrap(err, "destroying tenant")
 	}
-	if err := rejectIfSystemTenant(tenID, op); err != nil {
+
+	return destroyTenantInternal(ctx, p.txn, p.execCfg, &p.extendedEvalCtx, p.User(), info, synchronous)
+}
+
+// DestroyTenantByID implements the tree.TenantOperator interface.
+func (p *planner) DestroyTenantByID(ctx context.Context, tenID uint64, synchronous bool) error {
+	if err := p.validateDestroyTenant(ctx); err != nil {
 		return err
 	}
 
-	// Retrieve the tenant's info.
 	info, err := GetTenantRecordByID(ctx, p.execCfg, p.txn, roachpb.MustMakeTenantID(tenID))
 	if err != nil {
 		return errors.Wrap(err, "destroying tenant")
 	}
+	return destroyTenantInternal(ctx, p.txn, p.execCfg, &p.extendedEvalCtx, p.User(), info, synchronous)
+}
+
+func (p *planner) validateDestroyTenant(ctx context.Context) error {
+	const op = "destroy"
+	if err := p.RequireAdminRole(ctx, "destroy tenant"); err != nil {
+		return err
+	}
+	return rejectIfCantCoordinateMultiTenancy(p.execCfg.Codec, op)
+}
+
+func destroyTenantInternal(
+	ctx context.Context,
+	txn *kv.Txn,
+	execCfg *ExecutorConfig,
+	extendedEvalCtx *extendedEvalContext,
+	user username.SQLUsername,
+	info *descpb.TenantInfo,
+	synchronous bool,
+) error {
+	const op = "destroy"
+	tenID := info.ID
+	if err := rejectIfSystemTenant(tenID, op); err != nil {
```

Comment:
nit: should we move this check to `validateDestroyTenant`?

Reply:
Maybe there is a better way to factor things. But validateDestroyTenant are the checks we want to do before doing any work at all on behalf of the caller.  When dealing with names we need to do the lookup before we have the tenant ID, which is why I put this here. I'm going to leave it for now, but will revisit it when we come back to these functions to add the job cancellation stuff.